### DISCUSSION TITLE
How to package and install a custom Seq App

### DISCUSSION THREAD

**Discussion by gagwithgaffer** [datalust/seq-tickets#1140] [Help] [answered]
Hello, 

Last year Nick from your team was very helpful in putting together a custom app called 'Property Value List' and explained that it would be straight forward to take the code from there and produce my own app if I wanted to make some modifications.

I've decided that rather than writing the distinct property values to a text file that is stored somewhere within the Seq file system, I'd prefer to send each distinct value off to my own Web API so I can store the property names in a database.

I've prepared my own app but I just don't understand what is involved in order to package it up and install it into Seq for testing. Please note I'm only 2yrs old in my knowledge of programming and therefore have no experience uploading/posting/packaging my own code via GitHub or Nuget. Really looking for some guidance as a beginner in this area.

The class I've written is noted below:

```
namespace Seq.App.DistinctPropertyValues
{
    [SeqApp("Distinct Property Values", Description = "Track distinct values of a particular event property.")]
    class ValueListApp : SeqApp, ISubscribeTo<LogEventData>
    {
        readonly HashSet<string> _values = new HashSet<string>();

        [SeqAppSetting(
            DisplayName = "Property name",
            HelpText = "The name of the event property to track. Only non-whitespace string values are considered.")]
        public string PropertyName { get; set; }

        [SeqAppSetting(
            DisplayName = "Web API Base Uri",
            HelpText = "The Web API's base address. Example = http://localhost:8001/seq/createproperty")]
        public string WebApiUri { get; set; }

        public void On(Event<LogEventData> evt)
        {
            if (evt == null) throw new ArgumentNullException(nameof(evt));

            if (!evt.Data.Properties.TryGetValue(PropertyName, out var value) ||
                !(value is string s) ||
                string.IsNullOrWhiteSpace(s))
            {
                return;
            }

            if (TryAdd(s))
            {
                try
                {
                    using (WebClient webClient = new WebClient())
                    {
                        webClient.BaseAddress = WebApiUri;
                        // Optionally specify an encoding for uploading and downloading strings.
                        webClient.Encoding = Encoding.UTF8;
                        // Upload the data.
                        string reply = webClient.UploadString(webClient.BaseAddress, s);
                        Log.Information("New value {NewValue} detected for property {PropertyName}. Property value succesfully sent to an external Web API with response {message}", s, PropertyName, reply);

                    }
                }
                catch (WebException ex)
                {
                    String responseFromServer = ex.Message.ToString() + " ";
                    if (ex.Response != null)
                    {
                        using (WebResponse response = ex.Response)
                        {
                            Stream dataRs = response.GetResponseStream();
                            using (StreamReader reader = new StreamReader(dataRs))
                            {
                                responseFromServer += reader.ReadToEnd();
                            }
                        }
                    }
                    Log.Error("Distinct Property Value Web API returend an Error with a Server Response: " + responseFromServer);
                }
            }
        }

        public bool TryAdd(string item)
        {
            if (string.IsNullOrEmpty(item))
                throw new ArgumentException("Item cannot be null or empty.", nameof(item));

            if (_values.Contains(item))
                return false;

            _values.Add(item);
            return true;
        }
    }
}
```
![VS Solution](https://user-images.githubusercontent.com/51385414/108436388-402d0680-7243-11eb-938a-8553afd2c2d8.JPG)

**Comment by nblumhardt** [comment]
Hi! 

If your project follows the same structure as: https://github.com/datalust/seq-app-valuelist, then you can use the same Build.ps1 file that's in that repo in order to package the app.

Running `./Build.ps1` should create a package under `artifacts/` in the repository root.

The easiest way to test the package locally is to go to Settings > Feeds in Seq, add the `artifacts/` folder as a feed (you'll need its full path) and install from there.

Note you will have to set a different version number each time you package the app; the Build.ps1 script uses APPVEYOR_BUILD_NUMBER for this:

```
$env.APPVEYOR_BUILD_NUMBER = 1
./Build.ps1
```

Let me know if this helps!

Best regards,
Nick

**Comment by gagwithgaffer** [comment]


**Reply by gagwithgaffer** [reply]
Hi Nick, I appreciate your notes above but I really don't understand the process you describe. So trying to be self sufficient and persevere on, I later found another article which gives more information here: 
https://blog.datalust.co/building-modern-seq-apps-in-c-outputs/

So I build the app and then run the 'dotnet pack' command in the VS Package Manager Console. Details below:

![image](https://user-images.githubusercontent.com/51385414/108553022-03681a80-72ea-11eb-882c-766c52e6ddf9.png)

I then create a feed in Seq:

![image](https://user-images.githubusercontent.com/51385414/108553271-59d55900-72ea-11eb-993a-ee03883aec0c.png)

When I try to upload the app from my computer's local drive, I get the error warning below: Does this have anything to do with the fact that Seq is currently running in docker for Windows? the file path syntax looks the same as in your article, thx

![image](https://user-images.githubusercontent.com/51385414/108553388-7bcedb80-72ea-11eb-8677-6563ab34095a.png)

**Reply by nblumhardt** [reply]
Hi! Yes, this will be because you're running Dockerized Seq - on Windows, Docker containers run in a separate virtual machine with a different filesystem, so the code running in the container can't see C:\.

The simplest way to get past this will be to `docker stop` and `docker rm` the Seq container, then install Seq using the Windows MSI.

If you're comfortable with how volumes work under Docker, however, you can alternatively mount a folder from your local filesystem into the container, e.g. `-v C:\SeqApps\DistinctPropertyValues\bin\Debug:/feed` on the `docker run` command-line would mount your folder to `/feed` in the container, and you'd use `/feed` as the path when setting up the NuGet feed in Seq. Getting volume mounts to work between Windows and Docker can take some fiddling around, though, so I'd recommend going with the Windows installer if you have that option.

Hope this helps!
Nick

**Comment by gagwithgaffer** [comment, marked as answer]
Hi Nick,

Some good news I finally figured out the solution. Took me a number of days with trial and error learning the quirks of the process. To help others, I've spent most of today writing up a detailed article covering all the steps I went through and have posted it in the Show and tell category, hope you like it :)

[How to create a custom App for Seq - In Layman's Terms](https://github.com/datalust/seq-tickets/discussions/1142)

I'd be most honoured if you might consider packaging/validating the custom app to something more official through your repository. Thanks

**Reply by nblumhardt** [reply]
That's so awesome, thanks @gagwithgaffer !!!
